src/random.rs

```diff
@@ -0,0 +1,722 @@
+use crate::prelude::IndexOp;
+use crate::{error::Exception, Array, ArrayElement, StreamOrDevice};
+use mach_sys::mach_time;
+use mlx_macros::default_device;
+use std::sync::{Mutex, OnceLock};
+
+struct RandomState {
+    state: Array,
+}
+
+impl RandomState {
+    fn new() -> Self {
+        let now = unsafe { mach_time::mach_approximate_time() };
+        Self { state: key(now) }
+    }
+
+    fn next(&mut self) -> Array {
+        let next = split(&self.state);
+        self.state = next.0;
+        next.1
+    }
+
+    fn seed(&mut self, seed: u64) {
+        self.state = key(seed);
+    }
+}
+
+fn state() -> &'static Mutex<RandomState> {
+    static STATE: OnceLock<Mutex<RandomState>> = OnceLock::new();
+    STATE.get_or_init(|| Mutex::new(RandomState::new()))
+}
+
+/// Seed the random number generator.
+pub fn seed(seed: u64) {
+    let mut state = state().lock().unwrap();
+    state.seed(seed);
+}
+
+/// Get a PRNG key from a seed.
+///
+/// Return a value that can be used as a PRNG key.  All ``random::*``
+/// functions take an optional key -- this will let you control the
+/// random number generation.
+pub fn key(seed: u64) -> Array {
+    unsafe { Array::from_ptr(mlx_sys::mlx_random_key(seed)) }
+}
+
+/// Split a PRNG key into two keys and return a tuple.
+#[default_device]
+pub fn split_device(key: &Array, stream: StreamOrDevice) -> (Array, Array) {
+    let keys = unsafe {
+        Array::from_ptr(mlx_sys::mlx_random_split_equal_parts(
+            key.as_ptr(),
+            2,
+            stream.as_ptr(),
+        ))
+    };
+
+    (keys.index(0), keys.index(1))
+}
+
+/// Generate uniformly distributed random numbers with a given interval (`lower_bound:` and `upper_bound:`).
+///
+/// The values are sampled uniformly in the range.  An optional shape can be used to broadcast into
+/// a larger array.  An optional `key` can be specified to control the PRNG.
+///
+/// ```rust
+/// let key = mlx_rs::random::key(0);
+///
+/// // create an array of shape `[50]` type Float values in the range `0 ..< 10`
+/// let array = mlx_rs::random::uniform::<_, f32>(0, 10, &[50], &key);
+///
+/// // same, but in range `0.5 ..< 1`
+/// let array = mlx_rs::random::uniform::<_, f32>(0.5f32, 1f32, &[50], &key);
+/// ```
+#[default_device]
+pub fn uniform_device<'a, E: Into<Array>, T: ArrayElement>(
+    lower_bound: E,
+    upper_bound: E,
+    shape: &[i32],
```

Comment:
Is the shape optional?

Reply:
It is in python so I'll make it that way here too.